CPP: add a query for catching alloca in a loop

evverx opened the pull request:
Thanks to Sam Lanning (@samlanning) and Robert Marsh (@rdmarsh2) for taking the time to help
to make it possible. In fact, it was Robert Marsh who effectively
wrote the query and figured out that __builtin_alloca should be
used to also take functions like strdupa into account. I just
filled out the metadata :-)

@samlanning is my understanding correct that we kind of agreed that the query is perfect and self-documenting so there's no need for me to add tests or documentation? :-)

semmle-cla commented:
[![CLA assistant check](https://cla.semmle.com/pull/badge/not_signed)](https://cla.semmle.com/Semmle/ql?pullRequest=965) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.semmle.com/Semmle/ql?pullRequest=965) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.semmle.com/check/Semmle/ql?pullRequest=965) it.</sub>

evverx commented:
Forgot to add a link to the query in action: https://lgtm.com/query/6338234238766417258/

samlanning commented:
> @samlanning is my understanding correct that we kind of agreed that the query is perfect and self-documenting so there's no need for me to add tests or documentation? :-)

Haha, don't know about perfect ;) I think what I meant was that @rdmarsh2 offered to do the qhelp and unit tests. :)

evverx commented:
@geoffw0 thank you for the review! Fair enough, "high" was probably more about what I aspire to :-) I changed the precision to "medium" and updated the select as you suggested.

rdmarsh2 commented:
I'll let @geoffw0 finish up the review here and do tests and qhelp as a followup PR; @evverx do you mind if I ask for your review on that PR?

evverx commented:
@rdmarsh2 not at all. I'm not sure at this point I'll be of any help but I think it would be useful for me to take a look at how all the moving parts should be put together to complete the query. And thank you again for your help!

evverx commented:
Thank you!

How long does it usually take for a query to be rolled out on LGTM? If I understand correctly the results will be hidden by default once the query has reached LGTM. Would it be safe to include it in `.lgtm.yaml` as per https://lgtm.com/help/lgtm/showing-hiding-query-results right now?

sjvs commented:
There's a continuous testing and deployment (and data migration) cycle of the QL analysis engine and default queries on LGTM.com. Once a query has been merged, it'll become part of that cycle.

Yesterday we started running a smoke test for the next deployment: the next version of the analysis engine is being tested on a subset of the ~130,000 LGTM.com projects. Results will be double-checked, and a full test (on all 130k projects) will likely be run early next week, after which the new engine will be made live. 

Your query is unfortunately not part of the current cycle that started yesterday, which means it'll be part of the next cycle which should result in your query becoming live on LGTM.com in about two weeks.

Of course it's possible to run the query as a custom query on LGTM.com in the meantime; more info here: https://lgtm.com/help/lgtm/writing-custom-queries

aibaars commented:
@evverx It is safe to include the query in the `.lgtm.yml` file even though it does not exist yet on LGTM.com. Query `include` or `exclude` rules with query identifiers that do not (yet) exist have no effect and are simply ignored.
